FLOW-1619: Fix EMPTY operator for Objects and List iteration

ianhaycox opened the pull request:
Implemented IS_EMPTY for ContentObject

Re-factored getStateValue to move the GET_FIRST and GET_NEXT list iteration operations to Operations.ts. Also corrected an issue where for an Object property where getStateValue did not honour typeElementPropertyId and returned the whole object rather than the property.

Some fixes for the internaId/ExternalId problem when Offline and no externalId exists. This is to ensure the REMOVE functionality removes the correct item when Offline.

The ADD functionality is still broken when testing for duplicate inserts. This is again the internal/external ID problem. Working on that still.

jamesbrattBoomi commented on js/services/Rules.ts:
```diff
@@ -69,13 +69,16 @@ const Rules = {
 
             let left = snapshot.getValue(rule.leftValueElementToReferenceId);
-            left = getStateValue(rule.leftValueElementToReferenceId, left.typeElementId, left.contentType, '') || left;
+            // TODO - What to do if the operands have different contentTypes ? E.g.
```
I'm not 100% sure on this one. I think when trying to compare objects/lists with primitives, then an exception should probably be thrown. I'm not sure whether the expected behaviour is to use type coercion on comparing primitives or whether we support a deep diff on comparing objects.

ianhaycox replied:
Going to leave this one for now and tackle it later if it becomes an issue.  There's a lot of refactoring and error checking work to do, let alone work out the semantics of type conversions.

ianhaycox commented on js/models/State.ts:
```diff
@@ -28,39 +28,17 @@ export const StateInit = (state: IState) => {
- * @param command
  */
-export const getStateValue = (id: any, typeElementId: string, contentType: any, command: string) => {
+export const getStateValue = (id: any) => {
```
Done. Also added IFlow to some others

ianhaycox commented on js/models/State.ts:
```diff
@@ -28,39 +28,17 @@ export const StateInit = (state: IState) => {
-                if (property) {
-                    value.contentValue = property.contentValue ? property.contentValue : null;
-                    value.objectData = property.objectData ? property.objectData : null;
-                    value.index = 0;
```
The index isn't removed, it is used as the 'current pointer' to the array we are iterating with GET_NEXT, which is why it is saved in state so we can keep track of the loop.

ianhaycox commented on js/services/Operation.ts:
```diff
@@ -83,15 +83,27 @@ export const executeOperation = (operation: any, state: IState, snapshot: any) =
+            if (valueToReference.objectData) {
+                // Handle list iteration noting that GET_NEXT will get the first element if GET_FIRST has not been called beforehand
+                // Note that for iteration we need to persist the current array index by calling setStateValue()
+                if (manywho.utils.isEqual(operation.valueElementToReferenceId.command, 'GET_FIRST', true)) {
```
As discussed in Slack - it wasn't really any benefit to readability, so left as is.

ianhaycox commented:
Added tests for Rules
Added tests for Operations

Major linting exercise - sorry for all the changes

Fixed a minor issue with a missing space between the Caching^10% message
When creating a new value, the elementType is set to VARIABLE
Fixed SVG test warning
Fixed issue with 2 Reacts being loaded breaking hooks. Webpack configured to use external React and ReactDOM - same as ui-core
The linter raised a bunch of issue for unhandled Promises. In most cases I've just added a .catch((e) => console.error(e)), not ideal, but at least it won't fail silently now.

I ended up merging in Jose's state caching changes to ensure everything worked together and the appropriate linting was done on the merged code.

All tests run
Linting passes

Phew.

ianhaycox commented:
I can create a PR for the initial code fixes no problem. Adding the tests might be tricky, but I'll give it a go and maybe have to remove/skip the broken tests.  Should I create a new Jira issue for the new PR ?

jamesbrattBoomi commented:
Yeah, create a new Jira.

ianhaycox commented on js/components/NoNetwork.tsx:
```diff
@@ -3,22 +3,16 @@ import { INoNetworkProps } from '../interfaces/INoNetwork';
 
 declare const manywho: any;
```
Done

ianhaycox commented on js/components/Offline.tsx:
```diff
@@ -72,66 +62,77 @@ export class Offline extends React.Component<IOfflineProps, IOfflineState> {
+                        <span className="wait-message">
+                            { 'Caching ' }
+                            { String(this.props.cachingProgress) }
+                            %
```
Done

ianhaycox commented on js/components/Request.tsx:
```diff
@@ -78,40 +90,28 @@ class Request extends React.Component<IRequestProps, Partial<IRequestState>> {
+                extractExternalId(cachedRequest, tenantId, authenticationToken, stateId)
+                    .then(() => {
+                        this.onReplayResponse(resp);
+                    }, () => {});
```
OK

ianhaycox commented on js/components/Offline.tsx:
```diff
@@ -72,66 +62,77 @@ export class Offline extends React.Component<IOfflineProps, IOfflineState> {
+            this.props.toggleIsOffline({ hasNetwork: false });
+            this.props.toggleIsReplaying(false);
+            this.setState({ view: null });
+        });
```
Added

ianhaycox commented on js/components/GoOnline.tsx:
```diff
@@ -5,26 +5,23 @@ import { IGoOnlineProps, IGoOnlineState } from '../interfaces/IGoOnline';
-        },
-    };
-};
+const mapDispatchToProps = (dispatch) => ({
```
Changed as discussed
